### ISSUE TITLE
Support a fully qualified type name for arguments whose type is abstract

### ISSUE THREAD

**Issue by snoozbuster** [serilog/serilog-settings-configuration#71]
Recently, the `LiterateConsole` sink was deprecated, pending replacement by the normal `Console` sink. To maintain backward functionality, the `Console` sink now supports a `theme` parameter (see [themes](https://github.com/serilog/serilog-sinks-console)). This parameter is an instance of `ConsoleTheme`, an abstract class. I expected to be able to use the fully qualified type name as is possible for `IFormatter`, but it gives an error of "string not castable to ConsoleTheme" because apparently, the string-to-type name resolution is only supported [for interfaces](https://github.com/serilog/serilog-settings-configuration/blob/dev/src/Serilog.Settings.Configuration/Settings/Configuration/StringArgumentValue.cs#L57). Is there a reason for this, or could that line be extended to include abstract (and perhaps even concrete) class types?

**Comment by tsimbalar** [comment]
Hi,

Thanks for opening an issue :) 

You are correct, right now, the the *settings* part cannot handle this scenario for now. The *Settings* know how to call the *parameterless* constructor of a class, but because `ConsoleTheme` is abstract, it has no *callable* default constructor...  

There's a discussion about how we should approach it in here : serilog/serilog-sinks-console#16 , so I will close your issue as a duplicate, but feel free to add comments on that one :)

**Comment by tsimbalar** [comment]
Duplicate of serilog/serilog-sinks-console#16

**Comment by snoozbuster** [comment]
This is all quite true! But, `Serilog.Sinks.SystemConsole.Themes.EmptyConsoleTheme, Serilog.Sinks.Console, Version=3.1.0.0, Culture=neutral, PublicKeyToken=24c2f752a8e58a10` is a type which subclasses `ConsoleTheme` with a parameterless constructor, and I cannot see anything in the implementation of `StringArgumentValue.ConvertTo` which would break if that type was given in the settings. In general, I feel that there are use cases here and regardless of the solution to the theme problem (which is more complicated due to the implementation of `ConsoleTheme`), the restriction of `toTypeInfo.IsInterface` is too restrictive. Abstract classes can have subclasses with parameterless constructors too!

**Comment by tsimbalar** [comment]
My bad, I did not understand correctly and did not think it through ( replying issues before breakfast is a bad idea 🥐 ... 😃  ) 

You are correct, it seems rather logical that **`interface`s and `abstract class`es should be processed the same way**, and that the configuration should allow to specify a concrete implementation that has a default constructor by passing the fully qualified type name.

I guess the *Console* sink is the first one to introduce parameters whose type is an `abstract class` and that would explain why we missed it :) .

I'll reopen your issue for it is more general than the specific *console theme* issue.

AFAIK, such a change would need to be implemented both in : 
- Serilog *Core*'s handling of key-value settings (so that it is automatically available for *Serilog.Settings.AppSettings* , here : https://github.com/serilog/serilog/blob/3cd376c8c75ed49ab307c50cd80beaa06cbaecb7/src/Serilog/Settings/KeyValuePairs/SettingValueConversions.cs#L54-L71

- *Serilog.Settings.Configuration* , here : https://github.com/serilog/serilog-settings-configuration/blob/fae623c627496f00fcfab29f721cd49b2bbed989/src/Serilog.Settings.Configuration/Settings/Configuration/StringArgumentValue.cs#L57-L74


@serilog/reviewers-core : any opinion on this topic ?

**Comment by skomis-mm** [comment]
+1 to remove the restriction on interface type  👍 
For the console [issue](https://github.com/serilog/serilog-sinks-console/issues/16) I'm leaning towards to bring some kind of support to static readonly properties with proposed powershell-like syntax `Serilog.Sinks.SystemConsole.Themes.ConsoleTheme::None`

**Comment by nblumhardt** [comment]
👍 agreed on removing the restriction.

I'm also quite interested in the `::MemberName` syntax, if it's not to hard to parse reliably.

**Comment by tsimbalar** [comment]
@nblumhardt  I've created separate issues (serilog/serilog-settings-configuration#73 and serilog/serilog#1057 ) for the `::MemberName` part 

This issue can remain about **support type names for arguments whose type is abstract**, the same way we support interfaces ... and I think it makes sense to make it **up-for-grabs** ✨
